pkg/analysis/optionalfields/analyzer.go

```diff
@@ -190,7 +210,40 @@ func (a *analyzer) checkFieldPropertiesWithoutOmitEmpty(pass *analysis.Pass, fie
 		// Once it has the omitempty tag, it will also need to be a pointer in some cases.
 		// Now handle it as if it had the omitempty already.
 		// We already handle the omitempty tag above, so force the `hasOmitEmpty` to true.
-		a.checkFieldPropertiesWithOmitEmptyRequired(pass, field, fieldName, jsonTags, underlying, true, hasValidZeroValue, completeValidation, isPointer, isStruct)
+		a.checkFieldPropertiesWithOmitEmptyRequired(pass, field, fieldName, jsonTags, underlying, true, false, hasValidZeroValue, completeValidation, isPointer, isStruct)
+	}
+}
+
+func (a *analyzer) checkFieldPropertiesWithOmitZeroRequired(pass *analysis.Pass, field *ast.Field, fieldName string, jsonTags extractjsontags.FieldTagInfo, hasOmitZero, hasValidZeroValue, isPointer, isStruct bool) {
+	// In this case, we should always add the omitzero for struct fields if it isn't present.
+	a.handleFieldShouldHaveOmitZero(pass, field, fieldName, hasOmitZero, isStruct, jsonTags)
+
+	if !hasValidZeroValue && isStruct {
+		// The struct field need not be pointer if it does not have a valid zero value.
+		a.handleFieldShouldNotBePointer(pass, field, fieldName, isPointer, "field %s is optional and does not allow the zero value. The field does not need to be a pointer.")
+	}
+}
+
+func (a *analyzer) checkFieldPropertiesWithoutOmitZero(pass *analysis.Pass, field *ast.Field, fieldName string, jsonTags extractjsontags.FieldTagInfo, hasValidZeroValue, isPointer, isStruct bool) {
+	if !isStruct {
+		// Handle omitzero only for struct fields.
+		return
+	}
+
+	switch {
+	case hasValidZeroValue:
+		// The field is not omitzero, and the zero value is valid, the field does not need to be a pointer.
```

Comment:
This statement is wrong.

If the field has a valid zero value (ie `{}` is acceptable), then we want a pointer, with omitempty, but not omitzero